sympy/solvers/ode/systems.py

```diff
@@ -1064,20 +1065,234 @@ def canonical_odes(eqs, funcs, t):
+    for solution, derivatives in solutions_derivatives:
+        # Add parameter functions to handle underdetermined cases.
+        solution = _add_parameters_solve(solution, derivatives, t)
 
-    systems = []
-    for eq in canon_eqs:
-        system = [Eq(func.diff(t, order[func]), eq[func.diff(t, order[func])]) for func in funcs]
+        system = [Eq(lhs, rhs) for lhs, rhs in solution.items()]
         systems.append(system)
 
     return systems
 
 
+def _solve_reduce_derivatives(eqs, funcs, t):
+    #
+    # Solve for the highest derivatives of funcs in eqs.
+    #
+    # First try to solve for all derivatives of all orders. This will likely be
+    # an underdetermined system of equations but might give definite results
+    # for some derivatives. In particular we are interested in cases where we
+    # can solve for a lower order derivative than the highest derivative in the
+    # system. An example would be:
+    #
+    #   x' + y' = x
+    #   x' + y' = 0
+    #
+    # Here although the system appears to be first order in x (x' is the
+    # highest derivative) in fact it is zero order in x and we can conclude
+    # immediately that x=0 implying also that x'=0. The system is reduced to
+    #
+    #   y' = 0
+    #   x  = 0
+    #
+    from sympy.solvers.solvers import solve
+
+    order = _get_func_order(eqs, funcs)
+    highest_derivatives = [func.diff(t, order[func]) for func in funcs]
+
+    # There is no point in attempting any of the things below if the equations
+    # are already in explicit form. Maybe just use linsolve here?
+    derivs_found = []
+    derivs_set = set(highest_derivatives)
+    for eq in eqs:
+        if not eq.is_Equality:
+            break
+        for lhs, rhs in [(eq.lhs, eq.rhs), (eq.rhs, eq.lhs)]:
+            if lhs in derivs_set and not rhs.has_xfree(derivs_set):
+                derivs_found.append(eq.lhs)
+                break
+            # XXX: Catch a trivial Add or something else here...
+
+    if len(derivs_found) == len(eqs) == len(set(derivs_found)):
+        [sols] = solve(eqs, highest_derivatives, dict=True)
+        return [(sols, highest_derivatives)]
+
+    func2derivs = {}
+    deriv2func = {}
+
+    for f in funcs:
+        derivs = [f.diff(t, i) for i in range(order[f] + 1)]
+        func2derivs[f] = derivs
+        for n, dfdn in enumerate(derivs):
+            deriv2func[dfdn] = (f, n)
+
+    all_derivatives = list(deriv2func)
+
+    #
+    # Solve will solve somewhat randomly for different variables in the
+    # underdetermined system and returns effectively a new system of equations
+    # connecting different derivatives of different orders. What we want to do
+    # is for each solution find any equation that connects only derivatives
+    # that are *not* of the highest order. Such an equation can be used to
+    # eliminate higher order derivatives from the original equations.
+    #
+    solutions = solve(eqs, all_derivatives, dict=True)
```

Comment:
This can be very slow for some systems.

Reply:
In my opinion slow is still better that violating the equations provided by the user initially.

Reply:
Calling `solve` here is not necessary to prevent returning incorrect results: we can just reject invalid user input.

Reply:
I would not go as far as saying underdetermined input is "incorrect". Maybe the user is interested in just any representation of the potentially inifitely many solutions. But my viewpoint might be missing some general sympy primitives that would make it favorable to reject the input in this case. So I would leave the decision to you.

Reply:
It is not that underdetermined input is incorrect. It is that currently `dsolve_system` expects that all of the equations are differential equations. Equations that are not differential equations are invalid input and should be greeted with an error message rather than returning incorrect results.

Reply:
We could interpret the terms in these equations as the 0th derivative. Then we would not leave the realm of systems of odes.

Reply:
The example that you give in the comment is rather easy to reduce and see that actually no system of odes needs to be solved. But there might be cases where this is not so easiy to see. In those cases I would appreciate to still get a solution from sympy, instead of an error.

Reply:
I'm not sure which example you are referring to. It is better if you highlight the lines of code you are referring to.

Reply:
I am referring to lines 1092 to 1102
```python
...
#   x' + y' = x
#   x' + y' = 0
...
#   y' = 0
#   x  = 0
```

Reply:
The question is what is the general class of equations that is being handled and what is the general approach to reducing them. It is easy to come up with different methods that handle simple cases but we want a general method that does not rely on explicitly checking a large number of different special cases and is not wildly inefficient in other cases.

In this example we can do e.g.:
```python
In [16]: x, y = symbols('x, y', cls=Function)

In [17]: eqs = [x(t).diff(t) + y(t).diff(t) - x(t), x(t).diff(t) + y(t).diff(t)]

In [18]: solve(eqs, [x(t), y(t), x(t).diff(t), y(t).diff(t)], dict=True)
Out[18]: 
⎡⎧         d          d       ⎫⎤
⎢⎨x(t): 0, ──(x(t)): -──(y(t))⎬⎥
⎣⎩         dt         dt      ⎭⎦
```
Now we just work from the lowest derivatives upwards substituting into the highest derivatives.

The question is what happens when you apply the same techniques to other less trivial systems. Remember that such special cases are not actually the typical input. The typical input is a system of the form:
```
x' = f(x, y)
y' = g(x, y)
```
In this case it is never possible to do any algebraic reduction that eliminates any of the derivatives and the call to `solve` is always redundant.

Reply:
If I understood correctly, you fear that we loose performance for the main usecase of solving a 'proper' system of odes, only for allowing very few 'non-proper' systems of odes+daes to be solved as well.
In this case, I think we can only observe what other users say. For me it would be preferable to have it, as it allows me to solve my systems with less manual interaction, but when people notice their solves to take considerably more time, then we can just go back.
Nevertheless, that is the whole point of this PR, so I think we should at least give it a try.

Alternatively, we could introduce a keyword-argument to `sympy.dsolve` and `sympy.solvers.ode.systems.dsolve_system`, e.g. `try_reducing_dae=False`, which would skip this check by default and retain the performance of the current release (maybe also raise an error in case of 'invalid' input). Then the user has the chance to change it to `True` if desired as in my case.

Reply:
The issue is that this needs more careful testing with many more examples to consider exactly what is the best approach. Ideally we would not use `solve` here but something more well defined.

In particular solving for all derivatives like this is a bad idea if it leads to a polynomial system with many variables when the original equations were linear in the highest derivatives.

I don't remember what example I tested that showed this problem but there are definitely cases where that is not the right thing to do and so this code needs to be changed.

Reply:
What do you think about the keyword-argument suggested in my previous comment (`try_reducing_dae=False`)? That would allow the default to remain with the current implementation, and only try the additional solves if asked to.
This way users who give 'proper' systems of odes as inputs should not notice a difference to the current implementation. Or am I missing something?

Reply:
Keyword arguments like that are always a pain. There could be a separate function for transforming the system into an ordinary system of ODEs though.

Reply:
Ok. I am a little occupied this week, I will try to come up with something next week and let you know.

Reply:
> The question is what is the general class of equations that is being handled and what is the general approach to reducing them. It is easy to come up with different methods that handle simple cases but we want a general method that does not rely on explicitly checking a large number of different special cases and is not wildly inefficient in other cases.
> 
> In this example we can do e.g.:
> 
> ```python
> In [16]: x, y = symbols('x, y', cls=Function)
> 
> In [17]: eqs = [x(t).diff(t) + y(t).diff(t) - x(t), x(t).diff(t) + y(t).diff(t)]
> 
> In [18]: solve(eqs, [x(t), y(t), x(t).diff(t), y(t).diff(t)], dict=True)
> Out[18]: 
> ⎡⎧         d          d       ⎫⎤
> ⎢⎨x(t): 0, ──(x(t)): -──(y(t))⎬⎥
> ⎣⎩         dt         dt      ⎭⎦
> ```
> 
> Now we just work from the lowest derivatives upwards substituting into the highest derivatives.
> 
> The question is what happens when you apply the same techniques to other less trivial systems. Remember that such special cases are not actually the typical input. The typical input is a system of the form:
> 
> ```
> x' = f(x, y)
> y' = g(x, y)
> ```
> 
> In this case it is never possible to do any algebraic reduction that eliminates any of the derivatives and the call to `solve` is always redundant.

So I was thinking about this alot, and I came to agree with you, that this PR is missusing the ODE systems solver for systems that contain also other types of equations. Initially, I was not aware of this sharp distinction between ODEs and DAEs, but given the terminology and intention of the sympy ODE solver I see your point.
Nevertheless, I think it would be nice to have such functionality, as from applications often times these mixed systems of equations arise, and having a solution for them is of great interest.

> Keyword arguments like that are always a pain. There could be a separate function for transforming the system into an ordinary system of ODEs though.

I think this (transforming to 'proper' ODEs) is the most general approach. In agreement with you I think we should make it a separate function in a sense that it is called by the user instead of `dsolve` or `dsolve_system`, to not mess with the current usecase of 'proper' ODE systems. My suggestion is the following (names are of course working titles):

- new class in `sympy.solvers` called `mixed_dae_ode`
  - this class has a function called `solve_system`, which gets the user's system of equations. In general it does the following: 
    - identify DAEs in whole system, convert them to ODEs, submit 'new' whole system, which consists of 'proper' ODEs only, to `sympy.solvers.ode.systems.dsolve_system`
  - To identify the DAEs we can just check the order of the derivatives in an equation
  - For transforming the DAEs to ODEs I suggest a simple substitution of variables. After `dsolve_system` returns, we could just substitute back, and get the solution (if there is any) with the initial variable names.

Having this in a separate class allows to clearly state, e.g. in documentation, that this might lead to long runtimes, and if a system can be easily simplified by hand it might be worth doing it, instead of calling that function. Additionally, this approach would keep the ode namespace clean.

What do you think about this?

Reply:
I think the idea sounds good.

I don't think a class is needed though. If the class only has one method then it is really just a function.

The function should also transform initial conditions because there might be a need for a change of variables in the initial conditions.

Reply:
Good point about the initial conditions, didn't have them in mind.
My thoughts to introduce a class instead of a function were the following (no significance implied by order):

- clear distinction between existing 'proper' ODE functionality and new DAE+ODE functionality
- easier handling of  data-structures, e.g. substituted variables and their initial conditions. But I admit that this would still be possible with only a function
- during implementation there might come up some caveats for which additional functions come in handy. Then again they would be separated from the ODE namespace.